File: src/ios/AirshipLocationPlugin.m
/* Copyright Urban Airship and Contributors */

#import <Foundation/Foundation.h>
#import "AirshipLocationPlugin.h"
#import "UAirship.h"
#import <Cordova/CDVPlugin.h>
#import <CoreLocation/CoreLocation.h>
#import "UALocation.h"

typedef void (^UACordovaCompletionHandler)(CDVCommandStatus, id);
typedef void (^UACordovaExecutionBlock)(NSArray *args, UACordovaCompletionHandler completionHandler);

@interface AirshipLocationPlugin()
@property(nonatomic, strong) UALocation *location;
@end

@implementation AirshipLocationPlugin

- (void)pluginInitialize {
    UA_LINFO("Initializing Airship location cordova plugin.");

    self.location = [UALocation shared];
}

- (void)performCallbackWithCommand:(CDVInvokedUrlCommand *)command withBlock:(UACordovaExecutionBlock)block {
    [self.commandDelegate runInBackground:^{
        UACordovaCompletionHandler completionHandler = ^(CDVCommandStatus status, id value) {
            CDVPluginResult *result = [self pluginResultForValue:value status:status];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        };

        if (!block) {
            completionHandler(CDVCommandStatus_OK, nil);
        } else {
            block(command.arguments, completionHandler);
        }
    }];
}

/**
 * Helper method to create a plugin result with the specified value.
 *
 * @param value The result's value.
 * @param status The result's status.
 * @returns A CDVPluginResult with specified value.
 */
- (CDVPluginResult *)pluginResultForValue:(id)value status:(CDVCommandStatus)status {
    /*
     NSString -> String
     NSNumber --> (Integer | Double)
     NSArray --> Array
     NSDictionary --> Object
     NSNull --> no return value
     nil -> no return value
     */

    // String
    if ([value isKindOfClass:[NSString class]]) {
        NSCharacterSet *characters = [NSCharacterSet URLHostAllowedCharacterSet];
        return [CDVPluginResult resultWithStatus:status
                                 messageAsString:[value stringByAddingPercentEncodingWithAllowedCharacters:characters]];
    }

    // Number
    if ([value isKindOfClass:[NSNumber class]]) {
        CFNumberType numberType = CFNumberGetType((CFNumberRef)value);
        //note: underlyingly, BOOL values are typedefed as char
        if (numberType == kCFNumberIntType || numberType == kCFNumberCharType) {
            return [CDVPluginResult resultWithStatus:status messageAsInt:[value intValue]];
        } else  {
            return [CDVPluginResult resultWithStatus:status messageAsDouble:[value doubleValue]];
        }
    }

    // Array
    if ([value isKindOfClass:[NSArray class]]) {
        return [CDVPluginResult resultWithStatus:status messageAsArray:value];
    }

    // Object
    if ([value isKindOfClass:[NSDictionary class]]) {
        return [CDVPluginResult resultWithStatus:status messageAsDictionary:value];
    }

    // Null
    if ([value isKindOfClass:[NSNull class]]) {
        return [CDVPluginResult resultWithStatus:status];
    }

    // Nil
    if (!value) {
        return [CDVPluginResult resultWithStatus:status];
    }

    UA_LERR(@"Cordova callback block returned unrecognized type: %@", NSStringFromClass([value class]));
    return [CDVPluginResult resultWithStatus:status];
}

- (void)isLocationEnabled:(CDVInvokedUrlCommand *)command {
    UA_LTRACE("isLocationEnabled called with command arguments: %@", command.arguments);

    [self performCallbackWithCommand:command withBlock:^(NSArray *args, UACordovaCompletionHandler completionHandler) {
        BOOL enabled = self.location.locationUpdatesEnabled;
        completionHandler(CDVCommandStatus_OK, [NSNumber numberWithBool:enabled]);
    }];
}

- (void)isBackgroundLocationEnabled:(CDVInvokedUrlCommand *)command {
    UA_LTRACE("isBackgroundLocationEnabled called with command arguments: %@", command.arguments);

    [self performCallbackWithCommand:command withBlock:^(NSArray *args, UACordovaCompletionHandler completionHandler) {
        BOOL enabled = self.location.backgroundLocationUpdatesAllowed;
        completionHandler(CDVCommandStatus_OK, [NSNumber numberWithBool:enabled]);
    }];
}

- (void)setLocationEnabled:(CDVInvokedUrlCommand *)command {
    UA_LTRACE("setLocationEnabled called with command arguments: %@", command.arguments);

    [self performCallbackWithCommand:command withBlock:^(NSArray *args, UACordovaCompletionHandler completionHandler) {
        BOOL enabled = [[args objectAtIndex:0] boolValue];
        UA_LTRACE("setLocationEnabled set to:%@", enabled ? @"true" : @"false");

        dispatch_async(dispatch_get_main_queue(), ^{
            self.location.locationUpdatesEnabled = enabled;

            completionHandler(CDVCommandStatus_OK, nil);
        });

    }];
}

- (void)setBackgroundLocationEnabled:(CDVInvokedUrlCommand *)command {
    UA_LTRACE("setBackgroundLocationEnabled called with command arguments: %@", command.arguments);

    [self performCallbackWithCommand:command withBlock:^(NSArray *args, UACordovaCompletionHandler completionHandler) {
        BOOL enabled = [[args objectAtIndex:0] boolValue];

        dispatch_async(dispatch_get_main_queue(), ^{
            self.location.backgroundLocationUpdatesAllowed = enabled;

            completionHandler(CDVCommandStatus_OK, nil);
        });
    }];
}

@end

File: www/AirshipLocationPlugin.js
  /* Copyright Urban Airship and Contributors */
  
  var cordova = require("cordova"),
      exec = require("cordova/exec"),
      argscheck = require('cordova/argscheck');
  
  // Helper method to call into the native plugin
  function callNative(success, failure, name, args) {
    args = args || []
    exec(success, failure, "AirshipLocation", name, args)
  }
  
  /**
   * @module AirshipLocationPlugin
   */
  module.exports = {

    /**
     * Enables or disables Urban Airship location services.
     *
     * @param {Boolean} enabled true to enable location, false to disable.
     * @param {function} [success] Success callback.
     * @param {function(message)} [failure] Failure callback.
     * @param {string} failure.message The error message.
     */
    setLocationEnabled: function(enabled, success, failure) {
      argscheck.checkArgs('*FF', 'AirshipLocation.setLocationEnabled', arguments)
      callNative(success, failure, "setLocationEnabled", [!!enabled])
    },
  
    /**
     * Checks if location is enabled or not.
     *
     * @param {function(enabled)} success Success callback.
     * @param {boolean} success.enabled Flag indicating if location is enabled or not.
     * @param {function(message)} [failure] Failure callback.
     * @param {string} failure.message The error message.
     */
    isLocationEnabled: function(success, failure) {
      argscheck.checkArgs('fF', 'AirshipLocation.isLocationEnabled', arguments)
      callNative(success, failure, "isLocationEnabled")
    },
  
    /**
     * Enables or disables background location.
     *
     * @param {Boolean} enabled true to enable background location, false to disable.
     * @param {function} [success] Success callback.
     * @param {function(message)} [failure] Failure callback.
     * @param {string} failure.message The error message.
     */
    setBackgroundLocationEnabled: function(enabled, success, failure) {
      argscheck.checkArgs('*FF', 'AirshipLocation.setBackgroundLocationEnabled', arguments)
      callNative(success, failure, "setBackgroundLocationEnabled", [!!enabled])
    },
  
    /**
     * Checks if background location is enabled or not.
     *
     * @param {function(enabled)} success Success callback.
     * @param {boolean} success.enabled Flag indicating if background location updates are enabled or not.
     * @param {function(message)} [failure] Failure callback.
     * @param {string} failure.message The error message.
     */
    isBackgroundLocationEnabled: function(success, failure) {
      argscheck.checkArgs('fF', 'AirshipLocation.isBackgroundLocationEnabled', arguments)
      callNative(success, failure, "isBackgroundLocationEnabled")
    }

  }
  